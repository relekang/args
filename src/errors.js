// @flow
type CliErrorOptions = {
  message: string,
  exitCode: number,
  showHelp?: boolean
};
export class CliError extends Error {
  exitCode: number;
  showHelp: boolean;

  constructor({ message, exitCode, showHelp }: CliErrorOptions) {
    super(message);
    this.exitCode = exitCode;
    this.showHelp = showHelp || false;
  }

  toString() {
    return this.message;
  }
}
