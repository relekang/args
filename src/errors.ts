type CliErrorOptions = {
  message: string;
  exitCode: number;
  showHelp?: boolean;
};
export class CliError extends Error {
  exitCode: number;
  showHelp: boolean;

  constructor({ message, exitCode, showHelp }: CliErrorOptions) {
    super(message);
    this.exitCode = exitCode;
    this.showHelp = showHelp || false;

    Object.setPrototypeOf(this, CliError.prototype);
  }

  toString() {
    return this.message;
  }
}
