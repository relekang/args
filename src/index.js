// @flow

export type Config = {};

export default function args(_config: Config) {
  return ([_node, _program, ..._rest]: Array<String>) => ({});
}
