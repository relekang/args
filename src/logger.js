// @flow
/* eslint-disable no-console */

export function error(...args: *) {
  console.error(...args);
}

export function log(...args: *) {
  console.log(...args);
}
