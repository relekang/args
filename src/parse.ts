import mri from "mri";
import { CliError } from "./errors";
import { CommandConfig, CommandOption, Options } from "./types";

function evaluateOption(item: CommandOption, value: any) {
  if (item.required && !value) {
    throw new CliError({
      message: `Missing required argument "${item.name}"`,
      exitCode: 1,
      showHelp: true,
    });
  }
  value = item.transform ? item.transform(value) : value;
  if (!value) {
    return value;
  }
  if (typeof item.validate === 'function') {
    const error = item.validate(value);
    if (error) {
      throw new CliError({
        message: `Validation error on ${item.name}: ${error}`,
        exitCode: 1,
        showHelp: true,
      });
    }
  }
  return value;
}

export function parse(
  commandConfig: CommandConfig,
  args: Array<string>
): Options {
  const options: Options = mri(args);

  const positionals = (commandConfig.positionalOptions || []).reduce(
    (lastValue, item, index) => {
      const value = evaluateOption(item, options._[index]);
      if (value !== false && !value) {
        return lastValue;
      }
      return {
        ...lastValue,
        [item.name]: value,
      };
    },
    {}
  );

  const named = (commandConfig.namedOptions || []).reduce((lastValue, item) => {
    const value = evaluateOption(item, options[item.name]);
    if (value !== false && !value) {
      return lastValue;
    }
    return {
      ...lastValue,
      [item.name]: value,
    };
  }, {});

  return { ...options, ...positionals, ...named };
}
