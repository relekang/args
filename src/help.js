// @flow
/* eslint-disable global-require */
import chalk from "chalk";
import wrap from "word-wrap";
import path from "path";
import fs from "fs";
import { promisify } from "util";

import type { Config, CommandConfig, CommandOption, Options } from "./types";

import * as logger from "./logger";

const readdir = promisify(fs.readdir);

function values<A>(o: { [key: string]: A }): Array<A> {
  // $FlowFixMe
  return Object.values(o);
}

async function getCommands(config): { [key: string]: CommandConfig } {
  if (config.commands) {
    return config.commands;
  }
  return (await readdir(config.commandsPath)).filter(/\.js$/.test).reduce(
    (lastValue, file) => ({
      ...lastValue,
      // $FlowFixMe
      [file.replace(/.js$/, "")]: require(path.join(config.commandsPath, file))
    }),
    {}
  );
}

export async function createCommandsList(config: Config) {
  const commands = await getCommands(config);

  const length = Math.max(
    0,
    ...values(commands).map(({ name }) => name.length)
  );
  return values(commands)
    .map(command => {
      const spacing = new Array(Math.max(1, length - command.name.length + 1))
        .fill(" ")
        .join("");
      return `${command.name}${spacing} - ${command.help || ""}`;
    })
    .join("\n");
}

export function createPositionalUsage(command: CommandConfig) {
  return (command.positionalOptions || [])
    .map(({ name, required }) => (required ? `<${name}>` : `[<${name}>]`))
    .join(" ");
}

export function createUsageString(name: string, command: CommandConfig) {
  const positionals = createPositionalUsage(command);
  return (
    `Usage: ${name} ${command.name}` +
    (positionals ? ` ${positionals}` : "") +
    (command.namedOptions ? " [--options]" : "")
  );
}

export function createOptionHelpString(option: CommandOption) {
  return `  - ${option.name}${option.help ? ` - ${option.help}` : ""}`;
}

export function createSubCommandHelp(name: string, command: CommandConfig) {
  let lines = [
    chalk`{bold ${command.name}} - ${command.help} `,
    "\n",
    createUsageString(name, command),
    command.manual ? "\n" + wrap(command.manual, { width: 80 }) : ""
  ];
  if (command.positionalOptions) {
    lines = [...lines, "", chalk.bold("Arguments:")];
    lines = [
      ...lines,
      ...(command.positionalOptions || []).map(createOptionHelpString)
    ];
  }
  if (command.namedOptions) {
    lines = [...lines, "", chalk.bold("Options:")];
    lines = [
      ...lines,
      ...(command.namedOptions || []).map(createOptionHelpString)
    ];
  }
  return lines.join("\n");
}

function findSubCommand(config: Config, name: string): ?CommandConfig {
  if (config.commands) {
    // $FlowFixMe
    return config.commands[name];
  } else {
    try {
      // $FlowFixMe
      return require(path.join(config.commandsPath, name));
    } catch (error) {
      return;
    }
  }
}

export async function help(config: Config, options: Options) {
  const dashes = "--------------------------";
  const subCommand = findSubCommand(config, options._[0]);
  logger.error(chalk`{gray ${dashes}} {bold ${config.name}} {gray ${dashes}}`);
  logger.error("");
  if (subCommand) {
    logger.error(createSubCommandHelp(config.name, subCommand));
  } else {
    logger.error(chalk.bold`Commands: `);
    logger.error(await createCommandsList(config));
  }
}
