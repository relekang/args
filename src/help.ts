/* eslint-disable global-require */
import chalk from 'chalk';
import wrap from 'word-wrap';
import path from 'path';
import fs from 'fs';
import { promisify } from 'util';

import { Config, CommandConfig, CommandOption, Options } from "./types";

import * as logger from './logger';

const readdir = promisify(fs.readdir);

function values<A>(o: { [key: string]: A }): Array<A> {
  // $FlowFixMe
  return Object.values(o);
}

async function getCommands(
  config: Config
): Promise<{ [key: string]: CommandConfig }> {
  // @ts-ignore
  if (config.commands) {
    // @ts-ignore
    return config.commands;
  }
  // @ts-ignore
  return (await readdir(config.commandsPath))
    .filter(file => /\.[jt]s$/.test(file))
    .reduce(
      (lastValue, file) => ({
        ...lastValue,
        [file.replace(/.[jt]s$/, "")]: require(path.join(
          // @ts-ignore
          config.commandsPath,
          file
        )),
      }),
      {}
    );
}

export async function createCommandsList(config: Config) {
  const commands = await getCommands(config);

  const length = Math.max(
    0,
    ...values(commands).map(({ name }) => name.length)
  );
  return values(commands)
    .map(command => {
      const spacing = new Array(Math.max(1, length - command.name.length + 1))
        .fill(' ')
        .join('');
      return `${command.name}${spacing} - ${command.help || ''}`;
    })
    .join('\n');
}

export function createPositionalUsage(command: CommandConfig) {
  return (command.positionalOptions || [])
    .map(({ name, required }) => (required ? `<${name}>` : `[<${name}>]`))
    .join(' ');
}

export function createUsageString(name: string, command: CommandConfig) {
  const positionals = createPositionalUsage(command);
  return (
    `Usage: ${name} ${command.name}` +
    (positionals ? ` ${positionals}` : '') +
    (command.namedOptions ? ' [--options]' : '')
  );
}

export function createOptionHelpString(option: CommandOption) {
  return `  - ${option.name}${option.help ? ` - ${option.help}` : ''}`;
}

export function createSubCommandHelp(name: string, command: CommandConfig) {
  let lines = [
    chalk`{bold ${command.name}} - ${command.help} `,
    '\n',
    createUsageString(name, command),
    command.manual ? '\n' + wrap(command.manual, { width: 80 }) : '',
  ];
  if (command.positionalOptions) {
    lines = [...lines, '', chalk.bold('Arguments:')];
    lines = [
      ...lines,
      ...(command.positionalOptions || []).map(createOptionHelpString),
    ];
  }
  if (command.namedOptions) {
    lines = [...lines, '', chalk.bold('Options:')];
    lines = [
      ...lines,
      ...(command.namedOptions || []).map(createOptionHelpString),
    ];
  }
  return lines.join('\n');
}

function findSubCommand(
  config: Config,
  name: string
): CommandConfig | undefined {
  // @ts-ignore
  if (config.commands) {
    // @ts-ignore
    return config.commands[name];
  } else {
    try {
      // @ts-ignore
      return require(path.join(config.commandsPath, name));
    } catch (error) {
      return undefined;
    }
  }
}

export async function help(config: Config, options: Options) {
  const dashes = '--------------------------';
  const subCommand = findSubCommand(config, options._[0]);
  logger.error(chalk`{gray ${dashes}} {bold ${config.name}} {gray ${dashes}}`);
  logger.error('');
  if (subCommand) {
    logger.error(createSubCommandHelp(config.name, subCommand));
  } else {
    logger.error(chalk.bold`Commands: `);
    logger.error(await createCommandsList(config));
  }
}
