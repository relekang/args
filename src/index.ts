/* eslint-disable global-require */
import mri from 'mri';
import chalk from 'chalk';

import { help } from './help';
import * as logger from './logger';
import { CliError } from './errors';
import { parse } from './parse';
import { setup } from './setup';
import { Config, Options } from './types';
import { findSubCommand } from './subCommand';

export * from './errors';

async function __args(config: Config, subCommand: string, args: Array<string>) {
  await setup(config);
  if (subCommand === 'help') {
    return help(config, mri(args));
  }

  const command = findSubCommand(config, subCommand);

  if (command) {
    const options: Options = parse(command, args);
    await command.run(options);
  } else {
    throw new CliError({
      message: 'Unknown command',
      exitCode: 1,
      showHelp: true,
    });
  }
}

export function args(config: Config) {
  return ([_node, _program, subCommand, ...rest]: Array<string>) => {
    return __args(config, subCommand || config.defaultCommand || 'help', rest)
      .catch(async error => {
        if (
          error.constructor === CliError ||
          error.constructor.constructor === CliError
        ) {
          if (error.showHelp) {
            logger.error('\n');
            logger.error(chalk.bold(chalk.red(error.toString())));
            logger.error('\n\n');
            await help(config, mri(rest));
          } else {
            logger.error(error.toString());
          }
          if (process.env.NODE_ENV !== 'test') {
            process.exit(error.exitCode);
          }
        } else {
          logger.error(error);
        }
        if (process.env.NODE_ENV !== 'test') {
          process.exit(1);
        }
      })
      .catch(error => {
        logger.error(error.toString());
      });
  };
}

export default args;
