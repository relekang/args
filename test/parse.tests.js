// @flow
import { parse } from '../src/parse';
import type { CommandOption } from '../src/types';

const run = jest.fn();
const positionalOptions: Array<CommandOption> = [
  { name: 'first', required: true },
  { name: 'second', required: false, transform: Number },
  {
    name: 'third',
    required: false,
    validate: value => (value !== 't' ? "must be 't'" : null),
  },
];
const namedOptions: Array<CommandOption> = [
  { name: 'bool', required: false, transform: Boolean },
  { name: 'amount', required: false, transform: Number },
  { name: 'required', required: true },
  {
    name: 'validate',
    required: false,
    validate: value => (value !== 't' ? "must be 't'" : null),
  },
];

test('parse should parse positional options', () => {
  const parsed = parse({ name: 't', help: 't', positionalOptions, run }, [
    'first-arg',
    '200',
  ]);

  expect(parsed).toEqual({
    _: ['first-arg', '200'],
    first: 'first-arg',
    second: 200,
  });
});

test('parse should throw on missing required positional', () => {
  const fn = () => parse({ name: 't', help: 't', positionalOptions, run }, []);

  expect(fn).toThrowErrorMatchingSnapshot();
});

test('parse should throw on positional validation error', () => {
  const fn = () =>
    parse({ name: 't', help: 't', positionalOptions, run }, [
      'first',
      'second',
      'not-t',
    ]);

  expect(fn).toThrowErrorMatchingSnapshot();
});

test('parse should parse named options', () => {
  const parsed = parse(
    { name: 't', help: 't', namedOptions, run },
    '--required o/ --validate t'.split(' ')
  );

  expect(parsed).toEqual({
    _: [],
    validate: 't',
    bool: false,
    required: 'o/',
  });
});

test('parse should parse boolean named options', () => {
  const parser = args =>
    parse({ name: 't', help: 't', namedOptions, run }, args);

  expect(parser('--required o/ --bool'.split(' '))).toEqual({
    _: [],
    bool: true,
    required: 'o/',
  });

  expect(parser('--required o/ --no-bool'.split(' '))).toEqual({
    _: [],
    bool: false,
    required: 'o/',
  });
});

test('parse should throw on missing required named', () => {
  const fn = () => parse({ name: 't', help: 't', namedOptions, run }, []);

  expect(fn).toThrowErrorMatchingSnapshot();
});

test('parse should throw on named valdiation error', () => {
  const fn = () =>
    parse(
      { name: 't', help: 't', namedOptions, run },
      '--required o/ --validate woot'.split(' ')
    );

  expect(fn).toThrowErrorMatchingSnapshot();
});
