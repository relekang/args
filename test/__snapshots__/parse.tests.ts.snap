// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse should throw on missing required named 1`] = `"Missing required argument \\"required\\""`;

exports[`parse should throw on missing required positional 1`] = `"Missing required argument \\"first\\""`;

exports[`parse should throw on named valdiation error 1`] = `"Validation error on validate: must be 't'"`;

exports[`parse should throw on positional validation error 1`] = `"Validation error on third: must be 't'"`;
